### PR TITLE
Capture scope if necessary before Scala 2.13 Transformation enters an ExecutionContext

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

File: dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java [hunk 1]
```diff
@@ -77,6 +80,28 @@ private static void muzzleCheck(final Transformation callback) {
...
   }
 
+  public static final class Transfer {
+    @SuppressWarnings("rawtypes")
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static <F, T> void transfer(
+        @Advice.This Transformation<F, T> task,
+        @Advice.Argument(0) Object doNotUse,
+        @Advice.Argument(1) Object callbacks) {
+      if (callbacks instanceof Transformation && task != callbacks) {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java:90]: It seems like `callbacks` could be either `Transformation` or `ManyCallbacks`.  Would this still work if multiple callbacks were added to a single promise?

Comment (bantonsson) [reply dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java:90]: Nope, it won't work for multiple callbacks.

**Review by bantonsson — COMMENTED** [review]
I'm missing some rationale as to why this change is proposed. What is the use case that it's trying to fix?

**Comment by richardstartin** [comment]
@bantonsson the idea is when a `Transformation` is mapped, creating a new `Transformation`, any state from the first `Transformation` is passed on to the next one. Otherwise there may be no scope active when the mapped `Transformation` is constructed, so no span will be active when it's later run. Something like the `ConcurrentState` might be a better idea because we don't know that the first `Transformation` will never be run.

File: dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java (hunk 1 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java:90]: I don't think it's actually necessary to set the state in the first `Transformation` to null, and yes, it was racy.

Comment (richardstartin) [reply dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java:90]: It doesn't need to work with `ManyCallbacks` - see `Promise$Transformation`.

Comment (richardstartin) [reply dd-java-agent/instrumentation/scala-promise-2.13/src/main/java/datadog/trace/instrumentation/scala/concurrent/PromiseTransformationInstrumentation.java]: This instrumentation no longer exists

**Review by bantonsson — APPROVED** [review]

**Review by tylerbenson — APPROVED** [review]

**Pull request merged**